### PR TITLE
PyProject-centric Build System with Setuptools

### PR DESCRIPTION
**Pull request by ludgerpaehler**
This PR fully converts the existing setup tools-centric build system to a `pyproject.toml`-centric build system with [Poetry](https://python-poetry.org) as a part of this the PR introduces the following changes:

* Removal of `setup.cfg` with the flake8 configuration going its own dedicated `.flake8` configuration file
* Splitting up `requirements-dev.txt` into a `dev` dependency group, a `docs` dependency group to solely build the documentation, and moving `cvxpy` into the normal dependencies. If desired, `cvxpy` could be broken out into its own separate dependency group. These can be installed with `pip install pysindy[dev]`, or `pip install pysindy[docs]`.
* The workflows have been changed to accommodate the new build system
* The content of `setup.py` has been carried over into the `pyproject.toml`
* The README has been adjusted to the changes, and changed where necessary.

### CODE REVIEW

**Comment by Jacob-Stevens-Haas** [comment]
A PR to move from setup.py, requirements.txt, and setup.cfg to a pyproject.toml build has long been on the backburner, so a HUGE thanks for getting it started!  

That said, I'm more inclined to `setuptools_scm` than poetry as a backend, for a few reasons:
- It keeps metadata specification as in PEP 621 and PEP 631.  
- It auto-assigns version number from git tag (there is a 3rd party wrapper around poetry backend as an alternate)

Also, I'm not sure that committing a lockfile is appropriate, as we _want_ new versions of dependencies to cause CI failures, as that forces us to keep the library compatible with a wider range of dependencies.  I know some people have a workflow for package management that uses a lockfile, but I've never really understood them.

Was there some discussion about `poetry` that I missed?

**Comment by ludgerpaehler** [comment]
@Jacob-Stevens-Haas are you available to quickly triage the differences today/tomorrow? Started it during your talk on Monday, and just finished the first version yesterday.

**Comment by ludgerpaehler** [comment]
No personal preference between `setuptools`, and `poetry` tbh - am just more used to poetry so that is where I started, but happy to convert everything to `setuptools`, and utilize `setuptools_scm`.

As far as I can tell, the lock-file should be excluded from the PR - but if it is hiding somewhere, then I'll gladly rewrite the Git history to remove it.

I think the files where we already have full alignment are, and please correct me if I am wrong:
* `.flake8`
* `pre-commit-config.yaml`
* `.github/workflows/main.yml`
* `README.rst`

With most of the others being dependent on the build system used by `pyproject.toml`, and hence being blocked by the switch of the build system to `setuptools`.

What is your opinion of how the packages are split up into the three groups with the normal installation with `cvxpy`, `dev`, and `docs`? Should packages be moved into different groups?

While letting Poetry determine the version constraints, I also managed to remove duplicate package quotations inside of the requirements, and the dev-requirements which are already installed due to other packages.

File: pyproject.toml [hunk 1]
```diff
@@ -1,23 +1,65 @@
...
+pytest-lazy-fixtures = "^1.0.0"
+flake8-builtins-unleashed = "^1.3.1"
+jupyter = "^1.0.0"
+jupytext = "^1.14.5"
+pre-commit = "^3.3.1"
+hypothesis = "^6.75.2"
+jupyter-contrib-nbextensions = "^0.7.0"
 
-[tool.setuptools_scm]
-write_to = "pysindy/version.py"
```
Comment (Jacob-Stevens-Haas) [inline pyproject.toml:6]: keep

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -39,6 +39,8 @@ repos:
     -   id: codespell
         types_or: [python, rst, markdown]
         args: ["-L hist"]
+        additional_dependencies:
+          - tomli
```
Comment (Jacob-Stevens-Haas) [inline .pre-commit-config.yaml:43]: Good catch!  Thanks.

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Thanks again!  Most comments are about keeping the `setuptools` backend and `setuptools_scm` plugin, and how that changes pyproject.toml.  Currently an [open issue](https://github.com/python-poetry/poetry/issues/3332) to extend poetry support to PEP-defined format.

Thanks also for the documentation upgrade and removing duplicate packages!

> What is your opinion of how the packages are split up into the three groups with the normal installation with cvxpy, dev, and docs? Should packages be moved into different groups?

I think I see what you mean with `pip install pysindy[cvxpy]`.   Only concern is that, as an optional dependency, we need to guard imports from cvxpy.  We tried to do that with `pip install pysindy[miosr]`, but I'm not sure we were truly able to do it correctly.  I'm interested to hear whether @znicolaou  and @akaptano think that `cvxpy` is a difficult enough requirement to justify putting it in its own optional dependency. FWIW, MIOSR/`gurobipy`,  requires a paid (or academic) license to use the full functionality.

File: pyproject.toml (hunk 1 above)
Comment (Jacob-Stevens-Haas) [inline pyproject.toml:6]: Keep the `[tool.setuptools_scm]` table.

File: pyproject.toml [hunk 3]
```diff
@@ -1,23 +1,65 @@
...
+[tool.poetry.group.docs.dependencies]
+pandoc = "^2.3"
+sphinx-rtd-theme = "^1.2.0"
+sphinxcontrib-apidoc = "^0.3.0"
+nbsphinx = "^0.9.1"
```
[the remark this answers is not in the document]
Comment (ludgerpaehler) [reply pyproject.toml:50]: Poetry resolved the version this way, which might honestly also have some of its root in it resolving versioning on an M1-Mac, where versions tend to be much sparser, and much more heavily leaning towards newer Python versions.

[the remark this answers is not in the document]
Comment (ludgerpaehler) [reply pyproject.toml:50]: Poetry resolved the version this way, which might honestly also have some of its root in it resolving versioning on an M1-Mac, where versions tend to be much sparser, and much more heavily leaning towards newer Python versions.

**Comment by ludgerpaehler** [comment]
@Jacob-Stevens-Haas now the PR should be ready to merge :) The Readme is adjusted, and should be good-to-go.

Thanks for sitting down with me!

**Comment by Jacob-Stevens-Haas** [comment]
Thanks @ludgerpaehler !  I noticed that for some reason, codecov shows:
> Patch coverage has no change and project coverage change: -92.32 warning

which is odd, because we didn't change anything with pytest-coverage or codecov.  I'm going to merge it now and troubleshoot over the weekend, @akaptano @znicolaou , since codecov has been going through some weird changes recently.

**Pull request merged**